### PR TITLE
Mesh Free Problem Generator

### PR DESCRIPTION
**Pull request by sslattery**
Summary of PR:

1. Defines an interface to use for point cloud test problem generation. This includes parallel point coordinates, point ids, and field data.
2. Add an implementation of the generator for exodus files. The generator reads the exodus files, partitions the nodes of the elements in the mesh, and returns the problem.
3. Defines tests for the current mesh free operators using the exodus problem generator.

Notes:

- The problem generator builds both uniquely owned and ghosted sets of points with corresponding unique global ids.
- The unit test for `DataTransferKit::NearestNeighborOperator` only uses the unique partitioning of points as the API only takes uniquely partitioned points. The ghosted problem generator will be used when we write tests for the decorators that transition ghosted points to uniquely owned points.

To Do:

- [x] Add ghosted operator test.
- [x] Fix build dependency so generator tests only build when netcdf is available and other tests always build when tests are on.

### CODE REVIEW

File: packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+
+    // On each rank get the source coordinates with no partitioning and find
+    // the nearest neighbors to the locally-owned target points.
+    for ( int r = 0; r < comm->getSize(); ++r )
+    {
+        Kokkos::View<DataTransferKit::Coordinate **, Kokkos::LayoutLeft, Device>
+            src_coords;
+
+        // Load the source coordinates.
+        {
```
Comment (dalg24) [inline packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp]: Why do you have a curly brace around that bock?  Are you introducing a new scope on purpose to shadow the `tgt_coords` argument???  I was wondering how the call to `createUniquelyOwnedProblem()` even compiled since `tgt_coords` is a const reference in the declaration of `computeNeighborsBruteForce()`...

File: packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+//---------------------------------------------------------------------------//
+// Brute-force calculate the nearest neighbors
+template <class Device>
+void computeNeighborsBruteForce(
+    const Teuchos::RCP<const Teuchos::Comm<int>> &comm,
+    const std::string &src_exodus_file, const std::string &tgt_exodus_file,
+    const Kokkos::View<DataTransferKit::Coordinate **, Kokkos::LayoutLeft,
+                       Device> &tgt_coords,
+    Kokkos::View<DataTransferKit::Coordinate **, Kokkos::LayoutLeft,
+                 Kokkos::Serial> &src_nearest_coords )
```
Comment (dalg24) [inline packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp]: Why do you work from file names rather than from `src_coords` and `tgt_coords`?

File: packages/Meshfree/cmake/Dependencies.cmake [hunk 3]
```diff
@@ -5,4 +5,6 @@ TRIBITS_PACKAGE_DEFINE_DEPENDENCIES(
   Kokkos
   Teuchos
   Tpetra
+  TEST_REQUIRED_TPLS
```
[the remark this answers is not in the document]
Comment (sslattery) [reply packages/Meshfree/cmake/Dependencies.cmake]: Not quite sure on the CMake logic here. Lets talk about this tomorrow

File: packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator_def.hpp [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+        Kokkos::View<Scalar **, Kokkos::LayoutLeft, SourceDevice> &src_field,
+        Kokkos::View<Coordinate **, Kokkos::LayoutLeft, TargetDevice>
+            &tgt_coords,
+        Kokkos::View<Scalar **, Kokkos::LayoutLeft, TargetDevice> &tgt_field )
+{
+    // Partition the source coordinates in the x direction.
+    partitionUniquelyOwned( 0, _src_exodus_file, src_coords );
+
+    // Partition the target coordinates in the y direction.
+    partitionUniquelyOwned( 1, _tgt_exodus_file, tgt_coords );
```
[the remark this answers is not in the document]
Comment (sslattery) [reply packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator_def.hpp]: done

File: packages/Meshfree/test/PointCloudProblemGenerator/PointCloudProblemGenerator.hpp [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+     *
+     * \param tgt_coords Coordinates of the target points. Layout:
+     * (point,dim).
+     *
+     * \param tgt_field Field values at the target points. A field may have
+     * multiple components. Will be filled with zeros. Layout: (point,comp).
+     */
+    virtual void createUniquelyOwnedProblem(
+        const int field_dim,
+        Kokkos::View<Coordinate **, Kokkos::LayoutLeft, SourceDevice>
```
[the remark this answers is not in the document]
Comment (sslattery) [reply packages/Meshfree/test/PointCloudProblemGenerator/PointCloudProblemGenerator.hpp]: Two reasons. 1) If we wanted to switch to a `Tpetra::MultiVector` interface we would require views of `LayoutLeft` to directly construct vectors from the views. 2) Coordinates are stored in block contiguous format in the exodus files so it was easy to directly load the coordinates from the exodus file into the view.

File: packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp (hunk 2 above)
Comment (sslattery) [reply packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp]: To make it easy I just wanted to reload the coordinates in a serial partitioning on each rank. This obviously means doing some extra work but it was an easy way to get the coordinates.

File: packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp (hunk 1 above)
Comment (sslattery) [reply packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp]: yeah that is wrong shadowing is wrong - I have fixed this. The bracket scope is just so the dummy variables (who's names have been changed to reflect this) don't live after the loading of the exodus files.

File: packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    using DeviceType = typename Node::device_type;
+    using ExecutionSpace = typename DeviceType::execution_space;
+    using Scalar = double;
+
+    // Get the communicator.
+    auto comm = Teuchos::DefaultComm<int>::getComm();
+
+    // Create a problem generator.
+    std::string src_exodus_file = "fine_sphere.exo";
+    std::string tgt_exodus_file = "coarse_sphere.exo";
```
[the remark this answers is not in the document]
Comment (sslattery) [reply packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp]: done

File: packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+
+//---------------------------------------------------------------------------//
+
+// Include the test macros.
+#include "DataTransferKitMeshfree_ETIHelperMacros.h"
+
+// Create the test group
+#define UNIT_TEST_GROUP( NODE )                                                \
+    TEUCHOS_UNIT_TEST_TEMPLATE_1_INSTANT( ExodusProblemGenerator, one_to_one,  \
+                                          NODE )
```
[the remark this answers is not in the document]
Comment (sslattery) [reply packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp]: Not yet. `NearestNeighborOperator` only takes uniquely partitioned coordinate sets. When we have decorators written that handle general overlap partitioning we will use the ghosted problem generator.

File: packages/Meshfree/test/PointCloudProblemGenerator/PointCloudProblemGenerator.hpp [hunk 8]
```diff
@@ -55,6 +58,8 @@ class PointCloudProblemGenerator
      * \brief Create a general problem where points may exist on multiple
      * processors. Points have a unique global id.
      *
+     * \param field_dim The number of components in the field.
```
Comment (dalg24) [inline packages/Meshfree/test/PointCloudProblemGenerator/PointCloudProblemGenerator.hpp]: This input parameter does not make sense to me.  The user can infer the number of components from `src_field.extent( 0 )`.  How many component is up to the derived class.
As for `ExodusProblemGenerator`, I do understand that you need it there because you are not actually reading the fields from the files and you are just allocating memory and filling with zero.  But in that case you can always pass `field_dim` to the constructor.

File: packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp [hunk 9]
```diff
@@ -125,7 +126,16 @@ TEUCHOS_UNIT_TEST_TEMPLATE_1_DECL( ExodusProblemGenerator, one_to_one, Node )
     // Get the communicator.
     auto comm = Teuchos::DefaultComm<int>::getComm();
 
-    // Create a problem generator.
+    // Set an epsilon for the search. The meshes (see below) mesh a sphere of
+    // radius 10. An epsilon of 1.0 is needed to fully capture nearest
```
Comment (dalg24) [inline packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp]: I don't like "is needed".  Maybe "is sufficient in order to".

File: packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp [hunk 10]
```diff
@@ -188,11 +198,6 @@ TEUCHOS_UNIT_TEST_TEMPLATE_1_DECL( ExodusProblemGenerator, one_to_one, Node )
                           tgt_coord_copy );
     Kokkos::fence();
 
-    // Set an epsilon for the search.
-    auto const epsilon_default =
-        DataTransferKit::DistributedSearchTreeImpl<DeviceType>::epsilon;
-    DataTransferKit::DistributedSearchTreeImpl<DeviceType>::epsilon = 1.0;
-
```
Comment (dalg24) [inline packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp]: I actually like it better just before the operator.   Note that it is only used at construction time so you may set it back to the default value right after.

File: packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp [hunk 11]
```diff
@@ -205,8 +210,7 @@ TEUCHOS_UNIT_TEST_TEMPLATE_1_DECL( ExodusProblemGenerator, one_to_one, Node )
                                                          num_src );
     auto src_field_copy = KOKKOS_LAMBDA( const size_t i )
     {
-        for ( size_t n = 0; n < field_dim; ++n )
-            src_field_device( i ) = src_field( i, n );
+        src_field_device( i ) = src_field( i, 0 );
     };
```
Comment (dalg24) [inline packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp]: Also you do not use the lambda anywhere else you don't need to make an object, you can directly declare it in the `parallel_for()`.

File: packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp (hunk 2 above)
Comment (dalg24) [reply packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp]: I do not see how it is easier than passing the two point cloud since you have already parsed the files and then possibly use Tpetra import/export capability if you want to import the whole source vector on the current rank.

Comment (sslattery) [reply packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp]: Right - I can refactor to do the import/export

File: packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp [hunk 12]
```diff
@@ -0,0 +1,434 @@
...
+    auto comm = Teuchos::DefaultComm<int>::getComm();
+
+    // Create a problem generator. Two tet meshes of a sphere centered at
+    // (0,0,0) with a radius of 10 are used. The source mesh is represented by
+    // a fine tet mesh and the target mesh is represented by a coarse tet mesh.
+    std::string src_exodus_file = "fine_sphere.exo";
+    std::string tgt_exodus_file = "coarse_sphere.exo";
+    auto generator = std::make_shared<DataTransferKit::ExodusProblemGenerator<
+        Scalar, DeviceType, DeviceType>>( comm, src_exodus_file,
+                                          tgt_exodus_file );
```
Comment (dalg24) [inline packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp]: Why using a smart pointer here?

File: packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp [hunk 13]
```diff
@@ -0,0 +1,434 @@
...
+    Kokkos::View<DataTransferKit::Coordinate **, Device> src_coords_device(
+        "src_coords_device", num_src, space_dim );
+    auto src_coord_copy = KOKKOS_LAMBDA( const size_t i )
+    {
+        for ( size_t n = 0; n < space_dim; ++n )
+            src_coords_device( i, n ) = src_coords( i, n );
+    };
+    Kokkos::parallel_for( Kokkos::RangePolicy<ExecutionSpace>( 0, num_src ),
+                          src_coord_copy );
+    Kokkos::fence();
```
Comment (dalg24) [inline packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp]: Why wouldn't you just do `Kokkos::deep_copy( src_coords_device, src_coords )`?

File: packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp [hunk 14]
```diff
@@ -0,0 +1,434 @@
...
+    Kokkos::deep_copy( host_tgt_field, tgt_field_device );
+    double data_val;
+    for ( size_t t = 0; t < num_tgt; ++t )
+    {
+        data_val =
+            data_func( nearest_src_coords( t, 0 ), nearest_src_coords( t, 1 ),
+                       nearest_src_coords( t, 2 ), 0 );
+        test_result( t, 0 ) =
+            std::abs( host_tgt_field( t ) - data_val ) < 1.0e-12;
+    }
```
Comment (dalg24) [inline packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp]: Why wouldn't you just do your assertions in here?

File: packages/Meshfree/test/PointCloudProblemGenerator/PointCloudProblemGenerator.hpp [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+     *
+     * \param field_dim The number of components in the field.
+     *
+     * \param src_coords Coordinates of the source points. Layout:
+     * (point,dim).
+     *
+     * \param src_field Field values at the source points. A field may have
+     * multiple components. A generator implementation may choose to fill this
+     * with meaningful field data or may choose to initialize it to
+     * zero. Layout: (point,comp).
```
Comment (dalg24) [inline packages/Meshfree/test/PointCloudProblemGenerator/PointCloudProblemGenerator.hpp]: Documentation got out of sync.  Also, why did you remove the field values all together?  I recall in previous review only asking for `field_dim` parameter to be removed.

File: packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator_def.hpp [hunk 16]
```diff
@@ -0,0 +1,397 @@
...
+    // Get the node data.
+    Kokkos::View<Coordinate **, Kokkos::LayoutLeft, Kokkos::Serial>
+        export_coords;
+    Kokkos::View<GlobalOrdinal *, Kokkos::LayoutLeft, Kokkos::Serial>
+        export_gids;
+    getNodeDataFromFile( exodus_file, export_coords, export_gids );
+
+    // Figure out the min and max coordinates in the given dimension.
+    Coordinate dim_max, dim_min;
+    dimensionMinAndMax( dim, export_coords, dim_min, dim_max );
```
[the remark this answers is not in the document]
Comment (sslattery) [reply packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator_def.hpp]: It would be nice to move `DTK_DetailsUtils.hpp` into the `Utils` package

[the remark this answers is not in the document]
Comment (sslattery) [reply packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator_def.hpp]: In fact that makes sense to do within the context of #320

[the remark this answers is not in the document]
Comment (dalg24) [reply packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator_def.hpp]: > It would be nice to move `DTK_DetailsUtils.hpp` into the `Utils` package

Will do it at some point.

[the remark this answers is not in the document]
Comment (dalg24) [reply packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator_def.hpp]: > In fact that makes sense to do within the context of #320

What do you mean?

[the remark this answers is not in the document]
Comment (sslattery) [reply packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator_def.hpp]: I see the code in `DTK_DetailsUtils.hpp` as a nice set of general utilities for `Kokkos::View` objects. Although admittedly this is additional scope over the definition of #320

File: packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp (hunk 14 above)
Comment (sslattery) [reply packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp]: I agree but it doesn't compile:
```
[ 98%] Building CXX object DataTransferKit/packages/Meshfree/test/CMakeFiles/DataTransferKitMeshfree_NearestNeighborExodusGenerator.dir/tstNearestNeighborExodusGenerator.cpp.o
/scratch/source/trilinos/release/DataTransferKit/packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp(267): error: identifier "out" is undefined
          detected during instantiation of "void ExodusProblemGenerator_one_to_one_UnitTest<Node>::runUnitTestImpl(Teuchos::FancyOStream &, __nv_bool &) const [with Node=Kokkos_Compat_KokkosSerialWrapperNode]" 
(403): here

/scratch/source/trilinos/release/DataTransferKit/packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp(267): error: identifier "success" is undefined
          detected during instantiation of "void ExodusProblemGenerator_one_to_one_UnitTest<Node>::runUnitTestImpl(Teuchos::FancyOStream &, __nv_bool &) const [with Node=Kokkos_Compat_KokkosSerialWrapperNode]" 
(403): here

2 errors detected in the compilation of "/tmp/tmpxft_0000292c_00000000-7_tstNearestNeighborExodusGenerator.cpp1.ii".
DataTransferKit/packages/Meshfree/test/CMakeFiles/DataTransferKitMeshfree_NearestNeighborExodusGenerator.dir/build.make:62: recipe for target 'DataTransferKit/packages/Meshfree/test/CMakeFiles/DataTransferKitMeshfree_NearestNeighborExodusGenerator.dir/tstNearestNeighborExodusGenerator.cpp.o' failed
make[2]: *** [DataTransferKit/packages/Meshfree/test/CMakeFiles/DataTransferKitMeshfree_NearestNeighborExodusGenerator.dir/tstNearestNeighborExodusGenerator.cpp.o] Error 2
CMakeFiles/Makefile2:5514: recipe for target 'DataTransferKit/packages/Meshfree/test/CMakeFiles/DataTransferKitMeshfree_NearestNeighborExodusGenerator.dir/all' failed
make[1]: *** [DataTransferKit/packages/Meshfree/test/CMakeFiles/DataTransferKitMeshfree_NearestNeighborExodusGenerator.dir/all] Error 2
Makefile:160: recipe for target 'all' failed
make: *** [all] Error 2
```

Ultimately I think this is due to the fact that this is a templated unit test and the macros just can't figure it out.

Comment (dalg24) [reply packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp]: Add the `Teuchos::FancyOStream &out` and `bool &success` arguments to your function.

File: packages/Meshfree/test/PointCloudProblemGenerator/PointCloudProblemGenerator.hpp (hunk 15 above)
Comment (sslattery) [reply packages/Meshfree/test/PointCloudProblemGenerator/PointCloudProblemGenerator.hpp]: So the big issue with the field being returned by the generator is also providing the expected result on the target points. Do we return the expected target result? 

I would argue the correct result depends on the operator one uses. So, for example, the nearest operator correct result is different from a standard interpolation. 

On the opposite side, one could argue that these generator tests just check "how close" we are to some expected solution. The nearest neighbor test would then check the results relative to some large tolerance.

@dalg24 what do you think?

Comment (dalg24) [reply packages/Meshfree/test/PointCloudProblemGenerator/PointCloudProblemGenerator.hpp]: Generator should at the very least allocate memory and returns properly sized view for the target values.  When it is an option it should fill the view with expected data.  In the context of testing, then it will let us assert that the answer is close enough.

File: packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator.hpp [hunk 17]
```diff
@@ -81,24 +81,32 @@ class ExodusProblemGenerator
...
+        Kokkos::View<Scalar **, Kokkos::LayoutLeft, SourceDevice> &src_field,
         Kokkos::View<Coordinate **, Kokkos::LayoutLeft, TargetDevice>
-            &tgt_coords ) override;
+            &tgt_coords,
+        Kokkos::View<Scalar **, Kokkos::LayoutLeft, TargetDevice> &tgt_field )
+        override;
 
     // Create a general problem where points may exist on multiple
-    // processors. Both source and target fields are initialized to zero.
+    // processors. Both source and target fields have 1 component and are
```
Comment (dalg24) [inline packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator.hpp]: Why just one?  I thought exodus files could include solution files.

File: packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator.hpp [hunk 18]
```diff
@@ -0,0 +1,160 @@
+/****************************************************************************
+ * Copyright (c) 2012-2017 by the DataTransferKit authors                   *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the DataTransferKit library. DataTransferKit is     *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ ****************************************************************************/
```
Comment (dalg24) [inline packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator.hpp]: Please update headers as in #359

File: packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator.hpp (hunk 17 above)
Comment (sslattery) [reply packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator.hpp]: I wasn't aware of solution files.

In this case the basic exodus generator just reads a mesh and allocates a field of a single component as it has no internal solution state. Other implementations of the interface could return fields of multiple components. If we generate some exodus solution files in the future we could refactor this to return fields with dimensions that we get from the solutions.

File: packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator.hpp (hunk 18 above)
Comment (sslattery) [reply packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator.hpp]: good catch

Comment (sslattery) [reply packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator.hpp]: done

Comment (dalg24) [reply packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator.hpp]: You are missing the SPDX identifier

Comment (sslattery) [reply packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator.hpp]: ah right forgot that
